### PR TITLE
WIP: Add "One Host Scan" Feature

### PR DESCRIPTION
**Pull request by ARaimbekov**
One host scan, program launch example
 ./grinder.py -oh -target 8.8.8.8 -sk SHODAN_KEY -ci CENSYS_ID -cs CENSYS_SEC

### CODE REVIEW

**Comment by manmolecular** [comment]
Thanks.

First of all, we discussed it before, but I still need to remind you:
1. The name of the pull request must include the basic features, fixes, ideas, or improvements. Not "Check it please, i need some comments", it looks like a joke.
2. You need to resolve all the conflicts before creating the pull request, and this case is not the exception from this rule. So you need to fix it. 
3. You need to fix the author of the commits: it's not really required, but it's the rule of the good tone if one developer worked for some project on the same computer in the same appliance, it's better to keep the original author of the commits.

File: grinder/core.py [hunk 1]
```diff
@@ -129,6 +131,32 @@ def __init__(
...
         self.db = GrinderDatabase()
 
+    def onehost(self, host_address: str) -> List[dict]:
+
+        if self.shodan_api_key == "YOUR_DEFAULT_API_KEY":
+            print(f"│ Shodan key is not defined. Skip scan.")
+            print(f"└ ", end="")
+            return []
+
+        dnsResolve = 'https://api.shodan.io/dns/resolve?hostnames=' + host_address + '&key=' + self.shodan_api_key
```
Comment (manmolecular) [inline grinder/core.py:141]: Why are you using Shodan here? We can resolve DNS from the framework itself, with something like `host`, `whois`, any other utility, or maybe something even lighter.

Comment (manmolecular) [reply grinder/core.py:141]: Never use plus sign for string concatenation. Use format strings.

Comment (manmolecular) [reply grinder/core.py:141]: Naming convention - it's not JavaScript. Use underscore.

File: grinder/core.py [hunk 2]
```diff
@@ -129,6 +131,32 @@ def __init__(
...
+        resolved = requests.get(dnsResolve)
```
Comment (manmolecular) [inline grinder/core.py:142]: What if something goes wrong here? Where is the `try/except` block?

File: grinder/core.py [hunk 3]
```diff
@@ -129,6 +131,32 @@ def __init__(
...
+        host_address = resolved.json()[host_address]
```
Comment (manmolecular) [inline grinder/core.py:143]: What if response won't contain host_address? `IndexError`?

File: grinder/core.py [hunk 4]
```diff
@@ -129,6 +131,32 @@ def __init__(
...
+        shodan = ShodanConnector(api_key=self.shodan_api_key)
+        shodan.get_host_info(host_address)
+        shodan_raw_results = shodan.get_onehost_result()
+
+        with open("results_shodan.json", "w") as write_file:
+            json.dump(shodan_raw_results, write_file, indent=2)
```
Comment (manmolecular) [inline grinder/core.py:149]: First of all, we have a special module `filemanager.py` for these purposes, so you shouldn't use file logic directly in the core module, only in some special cases (this is not the special case, I suppose).

Comment (manmolecular) [reply grinder/core.py:149]: For what reason you use `indent=2`? Why `2`? Did you check other parts of the code? We use the value `4` everywhere.

File: grinder/core.py [hunk 5]
```diff
@@ -129,6 +131,32 @@ def __init__(
...
+
+        censys = CensysConnector(api_id=self.censys_api_id, api_secret=self.censys_api_secret)
+        censys.get_host_info(host_address)
+        censys_raw_results = censys.get_onehost_result()
```
Comment (manmolecular) [inline grinder/core.py:153]: Naming conventions, one_host

File: grinder/core.py [hunk 6]
```diff
@@ -129,6 +131,32 @@ def __init__(
...
+
+        with open("results_censys.json", "w") as write_file:
+            json.dump(censys_raw_results, write_file, indent=2)
```
Comment (manmolecular) [inline grinder/core.py:156]: The same comment about `filemanager.py` here.

Comment (manmolecular) [reply grinder/core.py:156]: Indent, again.

File: grinder/core.py [hunk 7]
```diff
@@ -149,8 +177,8 @@ def shodan_search(self, query: str, results_count: int = None) -> List[dict]:
 
         if not results_count:
             results_count = (
-                self.shodan_results_limit
-                or DefaultValues.SHODAN_DEFAULT_RESULTS_QUANTITY
+                    self.shodan_results_limit
+                    or DefaultValues.SHODAN_DEFAULT_RESULTS_QUANTITY
```
Comment (manmolecular) [inline grinder/core.py:181]: Check with Black linter.

File: grinder/core.py [hunk 8]
```diff
@@ -197,17 +225,17 @@ def censys_search(self, query: str, results_count: int = None) -> List[dict]:
 
         # Skip default values
         if (
-            self.censys_api_id == "YOUR_CENSYS_API_ID"
-            or self.censys_api_secret == "YOUR_CENSYS_API_SECRET"
+                self.censys_api_id == "YOUR_CENSYS_API_ID"
+                or self.censys_api_secret == "YOUR_CENSYS_API_SECRET"
```
Comment (manmolecular) [inline grinder/core.py:229]: Check with Black linter.

File: grinder/core.py [hunk 9]
```diff
@@ -197,17 +225,17 @@ def censys_search(self, query: str, results_count: int = None) -> List[dict]:
...
             print(f"│ Censys key is not defined. Skip scan.")
             print(f"└ ", end="")
             return []
 
         if not results_count:
             results_count = (
-                self.censys_results_limit
-                or DefaultValues.CENSYS_DEFAULT_RESULTS_QUANTITY
+                    self.censys_results_limit
+                    or DefaultValues.CENSYS_DEFAULT_RESULTS_QUANTITY
```
Comment (manmolecular) [inline grinder/core.py:238]: Check with Black linter.

File: grinder/core.py [hunk 10]
```diff
@@ -356,10 +384,10 @@ def count_vulnerabilities(self, max_vulnerabilities: int = 10) -> List[str]:
...
     @exception_handler(expected_exception=GrinderCoreLoadResultsFromFileError)
     def load_results_from_file(
-        self,
-        load_dir: str = DefaultValues.RESULTS_DIRECTORY,
-        load_file: str = DefaultValues.JSON_RESULTS_FILE,
-        load_json_dir: str = DefaultValues.JSON_RESULTS_DIRECTORY,
+            self,
+            load_dir: str = DefaultValues.RESULTS_DIRECTORY,
+            load_file: str = DefaultValues.JSON_RESULTS_FILE,
+            load_json_dir: str = DefaultValues.JSON_RESULTS_DIRECTORY,
```
Comment (manmolecular) [inline grinder/core.py:390]: Check with Black linter.

File: grinder/core.py [hunk 11]
```diff
@@ -703,7 +733,7 @@ def set_unique_entities_quantity(self, max_entities: int) -> None:
 
     @exception_handler(expected_exception=GrinderCoreCountUniqueProductsError)
     def count_unique_entities(
-        self, entity_name: str, search_results: dict = None, max_entities: int = None
+            self, entity_name: str, search_results: dict = None, max_entities: int = None
```
Comment (manmolecular) [inline grinder/core.py:736]: Check with Black linter.

File: grinder/core.py [hunk 12]
```diff
@@ -739,7 +769,7 @@ def count_unique_entities(
 
     @exception_handler(expected_exception=GrinderCoreHostShodanResultsError)
     def __parse_current_host_shodan_results(
-        self, current_host: dict, query: str, product_info: dict
+            self, current_host: dict, query: str, product_info: dict
```
Comment (manmolecular) [inline grinder/core.py:772]: Check with Black linter.

File: grinder/core.py [hunk 13]
```diff
@@ -751,8 +781,8 @@ def __parse_current_host_shodan_results(
         :return: None
         """
         if not (
-            current_host.get("location").get("latitude")
-            and current_host.get("location").get("longitude")
+                current_host.get("location").get("latitude")
+                and current_host.get("location").get("longitude")
```
Comment (manmolecular) [inline grinder/core.py:785]: Check with Black linter.

File: grinder/core.py [hunk 14]
```diff
@@ -781,7 +811,7 @@ def __parse_current_host_shodan_results(
 
     @exception_handler(expected_exception=GrinderCoreHostCensysResultsError)
     def __parse_current_host_censys_results(
-        self, current_host: dict, query: str, product_info: dict
+            self, current_host: dict, query: str, product_info: dict
```
Comment (manmolecular) [inline grinder/core.py:814]: Check with Black linter.

File: grinder/core.py [hunk 15]
```diff
@@ -1088,13 +1118,13 @@ def tls_scan(self, scanner_path: str) -> None:
...
-        host_timeout: int = DefaultNmapScanValues.HOST_TIMEOUT,
-        arguments: str = DefaultNmapScanValues.ARGUMENTS,
-        workers: int = DefaultNmapScanValues.WORKERS,
+            self,
+            ports: str = DefaultNmapScanValues.PORTS,
+            top_ports: int = DefaultNmapScanValues.TOP_PORTS,
+            sudo: bool = DefaultNmapScanValues.SUDO,
+            host_timeout: int = DefaultNmapScanValues.HOST_TIMEOUT,
+            arguments: str = DefaultNmapScanValues.ARGUMENTS,
+            workers: int = DefaultNmapScanValues.WORKERS,
```
Comment (manmolecular) [inline grinder/core.py:1127]: Check with Black linter.

File: grinder/core.py [hunk 16]
```diff
@@ -1144,13 +1174,13 @@ def nmap_scan(
...
-        workers: int = DefaultVulnersScanValues.WORKERS,
-        host_timeout: int = DefaultVulnersScanValues.HOST_TIMEOUT,
-        vulners_path: str = DefaultVulnersScanValues.VULNERS_SCRIPT_PATH,
+            self,
+            sudo: bool = DefaultVulnersScanValues.SUDO,
+            ports: str = DefaultVulnersScanValues.PORTS,
+            top_ports: int = DefaultVulnersScanValues.TOP_PORTS,
+            workers: int = DefaultVulnersScanValues.WORKERS,
+            host_timeout: int = DefaultVulnersScanValues.HOST_TIMEOUT,
+            vulners_path: str = DefaultVulnersScanValues.VULNERS_SCRIPT_PATH,
```
Comment (manmolecular) [inline grinder/core.py:1183]: Check with Black linter.

File: grinder/core.py [hunk 17]
```diff
@@ -1282,7 +1312,7 @@ def __filter_queries_by_vendor_confidence(self) -> None:
         self.queries_file = list(
             filter(
                 lambda product: str(product.get("vendor_confidence", "")).lower()
-                in required_confidences,
+                                in required_confidences,
```
Comment (manmolecular) [inline grinder/core.py:1315]: Check with Black linter.

File: grinder/core.py [hunk 18]
```diff
@@ -1322,7 +1352,7 @@ def __filter_queries_by_vendors(self) -> None:
         self.queries_file = list(
             filter(
                 lambda product: product.get("vendor").lower()
-                in map(str.lower, founded_vendors),
+                                in map(str.lower, founded_vendors),
```
Comment (manmolecular) [inline grinder/core.py:1355]: Check with Black linter.

File: grinder/core.py [hunk 19]
```diff
@@ -1354,8 +1384,8 @@ def run_scripts(self, queries_filename: str):
             scripts = None
             for product in self.queries_file:
                 if (product.get("vendor"), product.get("product")) == (
-                    host_info.get("vendor"),
-                    host_info.get("product"),
+                        host_info.get("vendor"),
+                        host_info.get("product"),
```
Comment (manmolecular) [inline grinder/core.py:1388]: Check with Black linter.

File: grinder/core.py [hunk 20]
```diff
@@ -1458,7 +1488,8 @@ def batch_search(self, queries_filename: str, not_incremental: bool = False) ->
 
         len_of_products = len(self.queries_file)
         for product_index, product_info in enumerate(self.queries_file):
-            cprint(f"{product_index} / {len_of_products} :: Current product: {product_info.get('product')}", "blue", attrs=["bold"])
+            cprint(f"{product_index} / {len_of_products} :: Current product: {product_info.get('product')}", "blue",
+                   attrs=["bold"])
```
Comment (manmolecular) [inline grinder/core.py:1492]: Check with Black linter.

File: grinder/defaultvalues.py [hunk 21]
```diff
@@ -10,6 +10,7 @@ class DefaultValues:
     CENSYS_API_ID: str = "YOUR_CENSYS_API_ID"
     CENSYS_API_SECRET: str = "YOUR_CENSYS_API_SECRET"
     VULNERS_API_KEY: str = "YOUR_VULNERS_API_KEY_HERE"
+    ONE_HOST_ADDRESS: str = "IP address"
```
Comment (manmolecular) [inline grinder/defaultvalues.py:13]: `IP_ADDRESS`

File: grinder/interface.py [hunk 22]
```diff
@@ -83,12 +83,24 @@ def parse_args(self) -> Namespace:
...
             help="JSON File with Shodan queries",
         )
+        parser.add_argument(
+            "-target",
+            "--host-ip",
+            action="store",
+            default=None,
+            help="IP address of host",
+        )
+
```
Comment (manmolecular) [inline grinder/interface.py:93]: Remove whitespace.

File: grinder/interface.py [hunk 23]
```diff
@@ -83,12 +83,24 @@ def parse_args(self) -> Namespace:
             default="queries.json",
             help="JSON File with Shodan queries",
         )
+        parser.add_argument(
+            "-target",
+            "--host-ip",
```
Comment (manmolecular) [inline grinder/interface.py:88]: Why the short form is `target`, but the full form is `--host-ip`? It's not logical.

File: grinder/interface.py [hunk 24]
```diff
@@ -83,12 +83,24 @@ def parse_args(self) -> Namespace:
...
+        )
+
         parser.add_argument(
             "-sk", "--shodan-key", action="store", default=None, help="Shodan API key"
         )
         parser.add_argument(
             "-vk", "--vulners-key", action="store", default=None, help="Vulners API key"
         )
+        parser.add_argument(
+            "-oh", "--onehost", action="store_true", default=False, help="One host scan"
```
Comment (manmolecular) [inline grinder/interface.py:101]: `--one-host`

File: grinder/interface.py [hunk 25]
```diff
@@ -83,12 +83,24 @@ def parse_args(self) -> Namespace:
...
+        )
+
```
Comment (manmolecular) [inline grinder/interface.py:103]: Remove whitespace.

File: grinder/shodanconnector.py [hunk 26]
```diff
@@ -13,11 +13,12 @@ class ShodanConnector:
     def __init__(self, api_key=DefaultValues.SHODAN_API_KEY):
         self.api = Shodan(api_key)
         self.results: list = []
+        self.onehost_result = {}
```
Comment (manmolecular) [inline grinder/shodanconnector.py:16]: Naming conventions, `one_host_result`

File: grinder/shodanconnector.py [hunk 27]
```diff
@@ -13,11 +13,12 @@ class ShodanConnector:
...
         self.shodan_results_count: int = 0
         self.real_results_count: int = 0
 
     def _remove_unused_fields_in_vulns(
-        self, max_references: int = DefaultValues.SHODAN_MAX_VULNERABILITIES_REFERENCES
+            self, max_references: int = DefaultValues.SHODAN_MAX_VULNERABILITIES_REFERENCES
```
Comment (manmolecular) [inline grinder/shodanconnector.py:21]: Check with Black linter.

File: grinder/shodanconnector.py [hunk 28]
```diff
@@ -30,14 +31,14 @@ def _remove_unused_fields_in_vulns(
             for cve, cve_information in host.get("vulns", {}).items():
                 if cve_information.get("references"):
                     cve_information["references"] = cve_information["references"][
-                        :max_references
-                    ]
+                                                    :max_references
+                                                    ]
```
Comment (manmolecular) [inline grinder/shodanconnector.py:35]: Check with Black linter.

File: grinder/shodanconnector.py [hunk 29]
```diff
@@ -30,14 +31,14 @@ def _remove_unused_fields_in_vulns(
...
                 if "verified" in cve_information.keys():
                     cve_information.pop("verified")
 
     @exception_handler(expected_exception=ShodanConnectorSearchError)
     def search(
-        self, query: str, max_records=DefaultValues.SHODAN_DEFAULT_RESULTS_QUANTITY
+            self, query: str, max_records=DefaultValues.SHODAN_DEFAULT_RESULTS_QUANTITY
```
Comment (manmolecular) [inline grinder/shodanconnector.py:41]: Check with Black linter.

File: grinder/shodanconnector.py [hunk 30]
```diff
@@ -84,3 +85,13 @@ def get_vulnerabilities(self) -> dict:
         return {
             host["ip_str"]: host["vulns"] for host in self.results if host.get("vulns")
         }
+
+    def get_onehost_result(self) -> dict:
```
Comment (manmolecular) [inline grinder/shodanconnector.py:89]: Docstring is missing.

File: grinder/shodanconnector.py [hunk 31]
```diff
@@ -84,3 +85,13 @@ def get_vulnerabilities(self) -> dict:
...
+        return self.onehost_result
+
+    def get_host_info(self, host_address: str):
```
Comment (manmolecular) [inline grinder/shodanconnector.py:92]: Docstring is missing.

File: grinder/interface.py [hunk 32]
```diff
@@ -83,12 +83,24 @@ def parse_args(self) -> Namespace:
             default="queries.json",
             help="JSON File with Shodan queries",
         )
+        parser.add_argument(
+            "-target",
```
Comment (manmolecular) [inline grinder/interface.py:87]: Do we really need this flag? We already have `-oh`, we can put our target in it.

**Comment by manmolecular** [comment]
@ARaimbekov @ParsevalBaks check the review.

File: grinder.py [hunk 33]
```diff
@@ -80,3 +80,5 @@
     if args.run:
         core.save_results_to_database()
     core.save_results()
+    if args.onehost:
```
[the remark this answers is not in the document]
Comment (ARaimbekov) [reply grinder.py:83]: Ok i will rewrite, thx for comment.

File: grinder.py [hunk 34]
```diff
@@ -80,3 +80,5 @@
...
+        core.onehost(host_address=args.host_ip)
```
[the remark this answers is not in the document]
Comment (ARaimbekov) [reply grinder.py:84]: Ok

File: grinder/censysconnector.py [hunk 35]
```diff
@@ -27,7 +27,9 @@ def __init__(
         self,
         api_id: str = DefaultValues.CENSYS_API_ID,
         api_secret: str = DefaultValues.CENSYS_API_SECRET,
+
```
[the remark this answers is not in the document]
Comment (ARaimbekov) [reply grinder/censysconnector.py:30]: Ok

File: grinder/censysconnector.py [hunk 36]
```diff
@@ -134,3 +136,13 @@ def get_results_count(self) -> int:
         :return: quantity of hosts
         """
         return self.censys_results_count
+    
+    def get_onehost_result(self) -> dict:
```
[the remark this answers is not in the document]
Comment (ARaimbekov) [reply grinder/censysconnector.py:140]: ok, i got it

File: grinder/censysconnector.py [hunk 37]
```diff
@@ -27,7 +27,9 @@ def __init__(
         self,
         api_id: str = DefaultValues.CENSYS_API_ID,
         api_secret: str = DefaultValues.CENSYS_API_SECRET,
+        
     ):
+        self.onehost_result = {}
```
[the remark this answers is not in the document]
Comment (ARaimbekov) [reply grinder/censysconnector.py:32]: ok, i got it

File: grinder/interface.py (hunk 32 above)
Comment (ARaimbekov) [reply grinder/interface.py:87]: good point

**Comment by manmolecular** [comment]
:warning: important update: please, when you will use Black for this PR, make sure that this changes will be in a separate commit, and this commit must be the last one, so I can check code and logic changes without any style differences, and after that, I can check PEP8 and other style compliance separately. Thanks.

And one more thing: please, if Black will "corrupt" all the code ("corrupt" in terms of rewriting all the module because of style, like, many other functions that are not related to your work at all), don't include it. I will format it by myself later in a separate PR. So, please, be accurate. Code reformatter of PyCharm and Black itself got kinda different style-logic, but Black is our solution and standard, so we use it. 

BTW, you can re-format dependencies and imports with PyCharm. But it's better not to use it for the code.

Also, if you want to skip some parts from formatting (if Black reformatting doesn't look right for you), you can use `fmt: off / fmt: on` tags. 

@ARaimbekov @ParsevalBaks

**Comment by manmolecular** [comment]
Any news? @ParsevalBaks @ARaimbekov

**Comment by manmolecular** [comment]
Do you plan to finish this PR? We still have **a lot** of open conversations here, and I don't see any real changes related to them. 

Also, as you can see, it's impossible to merge your PR, because you have conflicts with the original branch. 

Also, for god's sake, tell me, why did you upload the results of the scans (???) in the root directory of the project? 

As for now, this work is not done at all. It's even impossible to test it properly (moreover, you do not include any tests to do it). We can't even talk about coursework here.

@ARaimbekov @ParsevalBaks

**Comment by manmolecular** [comment]
What's new? 
I'm here to remind you that you have conflicts with the master branch, so you need to update your own branch + resolve all the conflicts. 

Also, I want to remind you, that it's better to fix it as early as possible (better for you).

**Comment by manmolecular** [comment]
???

**Comment by manmolecular** [comment]
Doesn't work now and won't be finished/fixed in the future obviously.

**Pull request closed without merging**